src/Npgsql/ConnectorPool.cs

```diff
@@ -197,200 +190,172 @@ internal async ValueTask<NpgsqlConnector> AllocateLong(NpgsqlConnection conn, Np
-                        {
-                            if (!waitingOpenAttempt.TaskCompletionSource.TrySetResult(null))
-                            {
-                                // TODO: Release more??
-                            }
-                        }
-
+                        Interlocked.Decrement(ref State.Open);
                         throw;
                     }
 
-                    Counters.NumberOfActiveConnections.Increment();
-                    Counters.NumberOfPooledConnections.Increment();
-
-                    // Start the pruning timer if we're above MinPoolSize
-                    if (_pruningTimer == null && newState.Total > _min)
+                    // We immediately store the connector as well, assigning it an index
+                    // that will be used during the lifetime of the connector for both _idle and _open.
+                    for (var i = 0; i < _open.Length; i++)
                     {
-                        var newPruningTimer = new Timer(PruningTimerCallback, this, -1, -1);
-                        if (Interlocked.CompareExchange(ref _pruningTimer, newPruningTimer, null) == null)
-                            newPruningTimer.Change(_pruningInterval, _pruningInterval);
-                        else
+                        if (Interlocked.CompareExchange(ref _open[i], connector, null) == null)
                         {
-                            // Someone beat us to it
-                            newPruningTimer.Dispose();
+                            connector.PoolIndex = i;
+                            break;
                         }
                     }
+                    Debug.Assert(connector.PoolIndex != int.MaxValue);
+
+                    // Only when we are the ones that incremented openCount past _min Change the timer.
+                    if (openCount - 1 == _min)
+                        _pruningTimer.Change(_pruningInterval, _pruningInterval);
+                    Counters.NumberOfPooledConnections.Increment();
+                    Counters.NumberOfActiveConnections.Increment();
+                    CheckInvariants(State);
 
                     return connector;
                 }
 
-                if (state.Busy == _max)
+                if (busyCount == _max)
                 {
-                    // Pool is exhausted. Increase the waiting count while atomically making sure the busy count
-                    // doesn't decrease (otherwise we have a new idle connector).
-                    checked
+                    // Pool is exhausted.
+                    // Enqueue an allocate attempt into the waiting queue so that the next release will unblock us.
+                    var tcs = new TaskCompletionSource<NpgsqlConnector?>(TaskCreationOptions.RunContinuationsAsynchronously);
+                    _waiting.Enqueue((tcs, async));
+
+                    // Scenario: pre-empted waiter
```

Comment:
Super nit: I think preempted and preemption are words (with the dash) :)

Reply:
I believe it's both correct, depending on english dialect it may either be preempt or pre-empt.